File: 2816_double_a_number_represented_as_a_linked_list/doubleIt_reverse.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverse(ListNode head) {
        ListNode tail = null;

        while (head != null) {
            ListNode currentNext = head.next;
            head.next = tail;
            tail = head;
            head = currentNext;
        }

        return tail;
    }

    public ListNode doubleIt(ListNode head) {
        ListNode current = reverse(head);
        ListNode tail = current;
        int carry = 0;


        while (current != null) {
            int value = current.val * 2;

            if (value >= 10) {
                value = value % 10;
                current.val = value + carry;
                carry = 1;
            } else {
                current.val = value + carry;
                carry = 0;
            }

            current = current.next;
        } 


        tail = reverse(tail);

        if (carry > 0) {
            tail = new ListNode(carry, tail);
        }

        return tail;
    }
}

File: 2816_double_a_number_represented_as_a_linked_list/doubleIt_reverse.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Double a Number Represented as a Linked List     
class Solution:
    def reverse(self, head: Optional[ListNode]) -> Optional[ListNode]:
        tail = None

        while head:
            currentNext = head.next
            head.next = tail
            tail = head
            head = currentNext

        return tail

    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current = tail = self.reverse(head)
        carry = 0
        
        while current:
            val = current.val * 2
            if val >= 10:
                val = val % 10
                current.val = val + carry
                carry = 1
            else:
                current.val = val + carry
                carry = 0
            current = current.next

        tail = self.reverse(tail)
        
        if carry:
            temp = tail
            tail = ListNode(carry)
            tail.next = temp

        return tail